File: 07-week-6--foundational-data-structures/03-bonus-algorithm--recursive-string-reverse/ruby/reverse_string.rb
def reverse_string(str)
  return str if str.length < 2

  return reverse_string(str[1..-1]) + str[0]
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'ih'"
  puts "=>", reverse_string('hi')

  puts

  puts "Expecting: 'ybabtac'"
  puts "=>", reverse_string('catbaby')

  # Don't forget to add your own!
end

# pseudocode to this file
# return string if length is 0 or 1
# return reverse_string(string - 0th character) + 0th character

# written explanation of solution
# Base case => we want to return if the string is empty or length == 1.

File: 07-week-6--foundational-data-structures/03-bonus-algorithm--recursive-string-reverse/javascript/reverse_string.js
function reverseString(str) {
  if (str.length < 2) {
    return str;
  } 
  return reverseString(str.slice(1)) + str[0];
}

if (require.main === module) {
  console.log("Expecting: 'ih'");
  console.log("=>", reverseString('ih'));

  console.log("");

  console.log("Expecting: 'ybabtac'");
  console.log("=>", reverseString('catbaby'));

  console.log("");

  console.log("Expecting: 'empty string'");
  console.log("=>", reverseString(''));

  console.log("");

  console.log("Expecting: 'esor'");
  console.log("=>", reverseString('rose'));
}

module.exports = reverseString;

// pseudocode
// return string if length is 0 or 1
// return reverse_string(string - 0th character) + 0th character

// written explanation of solution
// Base case => we want to return if the string is empty or length == 1.
